File: 2024/10 October/Delete Alternate Nodes/main.cpp
// Link: https://www.geeksforgeeks.org/problems/delete-alternate-nodes/1

#include <iostream>
#include <vector>
using namespace std;

struct Node
{
    int data;
    struct Node *next;

    Node(int x)
    {
        data = x;
        next = NULL;
    }
};

struct Node *createLinkedList(vector<int> &arr)
{
    struct Node *head = new Node(arr[0]);
    struct Node *tail = head;
    for (int i = 1; i < arr.size(); i++)
    {
        tail->next = new Node(arr[i]);
        tail = tail->next;
    }
    return head;
}

class Solution
{
public:
    void deleteAlt(struct Node *head)
    {
        Node *curr = head;
        while (curr && curr->next)
        {
            curr->next = curr->next->next;
            curr = curr->next;
        }
    }
};

int main()
{
    Solution obj;
    vector<int> arr = {1, 2, 3, 4, 5, 6};
    struct Node *head = createLinkedList(arr);
    obj.deleteAlt(head);
    while (head != NULL)
    {
        cout << head->data << " ";
        head = head->next;
    }
    cout << endl;

    arr = {99, 59, 42, 20};
    head = createLinkedList(arr);
    obj.deleteAlt(head);
    while (head != NULL)
    {
        cout << head->data << " ";
        head = head->next;
    }
    cout << endl;
    return 0;
}

File: 2024/10 October/Delete Alternate Nodes/main.py
# Link: https://www.geeksforgeeks.org/problems/delete-alternate-nodes/1


class Node:

    def __init__(self, data):
        self.data = data
        self.next = None


def printList(node):
    while node:
        print(node.data, end=" ")
        node = node.next
    print()


def createLinkedList(arr):
    head = Node(arr[0])
    last = head
    for val in arr[1:]:
        last.next = Node(val)
        last = last.next
    return head


class Solution:
    def deleteAlt(self, head):
        curr = head
        while curr and curr.next:
            curr.next = curr.next.next
            curr = curr.next


if __name__ == "__main__":
    obj = Solution()
    arr = [1, 2, 3, 4, 5, 6]
    head = createLinkedList(arr)
    obj.deleteAlt(head)
    while head is not None:
        print(head.data, end=" ")
        head = head.next
    print()

    arr = [99, 59, 42, 20]
    head = createLinkedList(arr)
    obj.deleteAlt(head)
    while head is not None:
        print(head.data, end=" ")
        head = head.next
    print()
